### PR TITLE
Basic client for async communication using 'json-rpc' package.

### PR DESCRIPTION
**Pull request by dmgav**
Implementation of basic async client that sends messages and processes responses using 'json-rpc' package. The features include timeout and basic handling of `json-rpc` errors. The functionality of the client is covered by unit tests.

The client is currently used for communication between RE Manager and Watchdog modules on the side of RE Manager. One of the next steps will be to convert communication between RE Manager and RE Worker to `json-rpc` using this client.

### CODE REVIEW

File: bluesky_queueserver/manager/comms.py [hunk 1]
```diff
@@ -1,13 +1,94 @@
...
+        # TODO: change 'code' and 'type' to read-only properties
+        self.__error_code__ = error_code
+        self.__error_type__ = error_type
+
+    @property
+    def error_code(self):
+        return self.__error_code__
+
+    @error_code.setter
+    def error_code(self, error_code):
```
Comment (tacaswell) [inline bluesky_queueserver/manager/comms.py:44]: Why over-ride the default `AttributeError` with a `RuntimeError`?

File: bluesky_queueserver/manager/comms.py [hunk 2]
```diff
@@ -102,3 +189,233 @@ def _conn_received(self, msg):
...
+
+    """
+    def __init__(self, conn, *, timeout=0.5, name="RE QServer Comm"):
+        self._conn = conn
+        self._loop = self._loop = asyncio.get_running_loop()
+
+        self._thread_name = name
+
+        self._fut_comm = None  # Future for waiting for messages from watchdog
+        self._event_comm = asyncio.Event()  # Event which is set when message is expected from watchdog
```
Comment (tacaswell) [inline bluesky_queueserver/manager/comms.py:251]: I don't think we need this Event.  We have the lock to make sure we only ever have one "send" in flight at a time and the event loop will make sure the async tasks don't step on each other's toes, and the `_expected_msg_id` attribute is enough to tell us "we are expecting a message back or not".

**Comment by tacaswell** [comment]
I pushed a few commits to (I hope) simplify the internal handshaking.

**Review by tacaswell — APPROVED** [review]

**Comment by tacaswell** [comment]
@dmgav If you are happy with my changes please merge.

File: bluesky_queueserver/manager/comms.py (hunk 1 above)
Comment (dmgav) [reply bluesky_queueserver/manager/comms.py:44]: Do you mean that we don't need the setter method here at all?

Comment (dmgav) [reply bluesky_queueserver/manager/comms.py:44]: I will merge now and do the changes in the next PR.

Comment (tacaswell) [reply bluesky_queueserver/manager/comms.py:44]: Yeah, if you don't provide a setter python provides a default failure mode

```
In [2]: class Bob: 
   ...:     @property 
   ...:     def foo(self): 
   ...:         return 1 
   ...: b = Bob()                                                                                                                                 

In [3]: b.foo = 5                                                                                                                                 
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
<ipython-input-3-b9a5dfa13086> in <module>
----> 1 b.foo = 5

AttributeError: can't set attribute
```

**Comment by dmgav** [comment]
I think changes make sense. Thanks @tacaswell

**Review by mrakitin — COMMENTED** [review]
I had a couple of questions/suggestions, but did not have time to submit them before the morning beamline work, and it got merged. I can submit a PR to follow up.

File: bluesky_queueserver/manager/comms.py [hunk 3]
```diff
@@ -102,3 +189,233 @@ def _conn_received(self, msg):
...
+                        return response["result"]
+                    elif "error" in response:
+                        # TODO: verify that this is all information that should be saved
+                        err_code = response["error"]["code"]
+                        if "data" in response["error"]:
+                            # Server Error (issue with execution of the method)
+                            err_type = response["error"]["data"]["type"]
+                            # Message: "Server error: <message text>"
+                            err_msg = response["error"]["message"] + ": " + \
+                                response["error"]["data"]["message"]
```
[the remark this answers is not in the document]
Comment (dmgav) [reply bluesky_queueserver/manager/comms.py:356]: I am not sure about using f-strings for combining other strings. Lately I prefer to add them. I am not sure what is right.

**Comment by dmgav** [comment]
@mrakitin I think it may be easier if I just fix the typos in the code used in the new PR.

File: bluesky_queueserver/manager/tests/test_comms.py [hunk 4]
```diff
@@ -2,17 +2,94 @@
...
+    assert count_threads_with_name(new_name) == 0, "No threads are expected to exist"
+
+    pc = PipeJsonRpcReceive(conn=conn2, name=new_name)
     pc.start()
+    assert count_threads_with_name(new_name) == 1, "One thread is expected to exist"
+
+    pc.start()  # Expected to do nothing
+
+    pc.stop()
+    ttime.sleep(0.15)  # Wait until the thread stops (0.1s polling period)
```
[the remark this answers is not in the document]
Comment (dmgav) [reply bluesky_queueserver/manager/tests/test_comms.py:88]: Polling period IS 0.1 s

File: bluesky_queueserver/manager/comms.py [hunk 5]
```diff
@@ -1,13 +1,94 @@
...
+    error_code: int
+        Error code (returned by `json-rpc`)
+    error_type: str
+        Error type (returned by `json-rpc` or set to `'CommJsonRpcError'`)
+    """
+    def __init__(self, message, error_code, error_type):
+        super().__init__(message)
+        # TODO: change 'code' and 'type' to read-only properties
+        self.__error_code__ = error_code
+        self.__error_type__ = error_type
```
[the remark this answers is not in the document]
Comment (dmgav) [reply bluesky_queueserver/manager/comms.py:37]: I guess it should be `__<>`. I'll change it.

File: bluesky_queueserver/manager/tests/test_comms.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (mrakitin) [reply bluesky_queueserver/manager/tests/test_comms.py:88]: Oh, I see. I'll fix it in the PR.

[the remark this answers is not in the document]
Comment (mrakitin) [reply bluesky_queueserver/manager/tests/test_comms.py:88]: Added a clarification and fixed via https://github.com/bluesky/bluesky-queueserver/pull/41/commits/295d8ede31d78d6b6c32cf639b0357498e9e81c2. I hope it is right now.

**Comment by mrakitin** [comment]
OK, please feel free to push to my PR branch #41 if you wish.

**Pull request merged**